File: ModInverse.java
long inv(long x , long mod){
	    long r, y;
	    for(r = 1 , y = mod - 2 ; y != 0 ;x = x * x % mod, y>>=1)
	    	if ((y & 1) == 1)
	    		r = r * x % mod;
	    return r;
	}

File: ModInverse.cpp
/*
(a / b) mod p = ((a mod p) * (b^(-1) mod p)) mod p
p = prime, b^(-1) mod p = b^(p - 2) mod p
*/

long long inv(long long x){
    long long r,y;
    for(r=1,y=M-2;y;x=x*x%M,y>>=1)(y&1)&&(r=r*x%M);
    return r;
}
